File: src/net/codechunk/speedofsound/service/SoundServiceManager.java
package net.codechunk.speedofsound.service;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.os.BatteryManager;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.util.Log;

/**
 * Sound service activation manager. Used to start the service at boot;
 * referenced in the manifest.
 */
public class SoundServiceManager extends BroadcastReceiver {
	private static final String TAG = "SoundServiceManager";

	public static final String LOCALE_BUNDLE = "com.twofortyfouram.locale.intent.extra.BUNDLE";
	public static final String LOCALE_BLURB = "com.twofortyfouram.locale.intent.extra.BLURB";
	public static final String LOCALE_FIRE = "com.twofortyfouram.locale.intent.action.FIRE_SETTING";

	/**
	 * Keep track of the bluetooth state here, as the undocumented broadcasts
	 * might not be sticky. Not sure if the official ones are either.
	 */
	private boolean bluetoothConnected = false;

	/**
	 * Get the filter of extra intents we care about.
	 */
	public IntentFilter activationIntents() {
		IntentFilter filter = new IntentFilter();

		// headset plug/unplug events (*mush* be registered dynamically)
		filter.addAction(Intent.ACTION_HEADSET_PLUG);

		// documented API11+ bluetooth API
		filter.addAction(android.bluetooth.BluetoothA2dp.ACTION_CONNECTION_STATE_CHANGED);

		return filter;
	}

	/**
	 * Receive a broadcast and start the service or update the tracking state.
	 */
	@Override
	public void onReceive(Context context, Intent intent) {
		String action = intent.getAction();
		Log.d(TAG, "Received intent " + action);

		// start the service on boot
		if (action.equals(Intent.ACTION_BOOT_COMPLETED)) {
			Intent startIntent = new Intent(context, SoundService.class);
			context.startService(startIntent);
			return;
		}

		// resume tracking if we're also in a satisfactory mode
		if (action.equals(Intent.ACTION_POWER_CONNECTED) ||
				action.equals(Intent.ACTION_POWER_DISCONNECTED) ||
				action.equals(Intent.ACTION_HEADSET_PLUG)) {
			SoundServiceManager.setTracking(context, this.shouldTrack(context));
		}

		// official API 11+ bluetooth A2DP broadcasts
		else if (action.equals(android.bluetooth.BluetoothA2dp.ACTION_CONNECTION_STATE_CHANGED)) {
			Log.d(TAG, "A2DP API11+ event");

			// set the bluetooth state
			int state = intent.getIntExtra(android.bluetooth.BluetoothA2dp.EXTRA_STATE, -1);
			if (state == android.bluetooth.BluetoothA2dp.STATE_CONNECTED) {
				Log.v(TAG, "A2DP active");
				this.bluetoothConnected = true;
			} else if (state == android.bluetooth.BluetoothA2dp.STATE_DISCONNECTED) {
				Log.v(TAG, "A2DP inactive");
				this.bluetoothConnected = false;
			}

			// start or stop tracking
			SoundServiceManager.setTracking(context, this.shouldTrack(context));
		}

		// locale/tasker intents
		else {
			if (action.equals(SoundServiceManager.LOCALE_FIRE)) {
				Bundle bundle = intent.getBundleExtra(SoundServiceManager.LOCALE_BUNDLE);
				if (bundle != null) {
					boolean state = bundle.getBoolean(SoundService.SET_TRACKING_STATE, true);
					SoundServiceManager.setTracking(context, state);
				}
			}
		}
	}

	/**
	 * Determine whether we should be tracking.
	 *
	 * @param context Application context
	 * @return suggested state
	 */
	private boolean shouldTrack(Context context) {
		// load preferences
		SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
		boolean powerPreference = prefs.getBoolean("enable_only_charging", false);
		boolean headphonePreference = prefs.getBoolean("enable_headphones", false);
		boolean bluetoothPreference = prefs.getBoolean("enable_bluetooth", false);

		// get power status
		IntentFilter plugFilter = new IntentFilter(Intent.ACTION_BATTERY_CHANGED);
		Intent powerStatus = context.getApplicationContext().registerReceiver(null, plugFilter);
		boolean powerConnected = false;
		if (powerStatus != null) {
			int plugState = powerStatus.getIntExtra(BatteryManager.EXTRA_PLUGGED, -1);
			powerConnected = (plugState == BatteryManager.BATTERY_PLUGGED_AC ||
					plugState == BatteryManager.BATTERY_PLUGGED_USB);
		} else {
			Log.e(TAG, "Power status was null");
		}

		// don't track if power is disconnected and we care
		if (powerPreference && !powerConnected) {
			Log.v(TAG, "Power preference active & disconnected");
			return false;
		}

		// get headphone status
		IntentFilter headsetFilter = new IntentFilter(Intent.ACTION_HEADSET_PLUG);
		Intent headphoneStatus = context.getApplicationContext().registerReceiver(null, headsetFilter);
		boolean headphoneConnected = false;
		if (headphoneStatus != null) {
			headphoneConnected = headphoneStatus.getIntExtra("state", 0) == 1;
		} else {
			Log.e(TAG, "Headphone status was null");
		}

		// activate if headphones are plugged in
		if (headphonePreference && headphoneConnected) {
			Log.v(TAG, "Headphone connected");
			return true;
		}

		// also activate if bluetooth is connected
		if (bluetoothPreference && this.bluetoothConnected) {
			Log.v(TAG, "Bluetooth connected");
			return true;
		}

		// anything else is a no-go
		return false;
	}

	/**
	 * Set the tracking state by sending a service start command.
	 *
	 * @param context Application context.
	 * @param state   Turn tracking on or off.
	 */
	private static void setTracking(Context context, boolean state) {
		Log.d(TAG, "Setting tracking state: " + state);
		Intent serviceIntent = new Intent(context, SoundService.class);
		serviceIntent.putExtra(SoundService.SET_TRACKING_STATE, state);
		context.startService(serviceIntent);
	}

}

File: src/net/codechunk/speedofsound/service/SoundServiceManager.kt
package net.codechunk.speedofsound.service

import android.bluetooth.BluetoothDevice
import android.content.*
import android.os.BatteryManager
import android.preference.PreferenceManager
import android.util.Log
import androidx.core.content.ContextCompat
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import net.codechunk.speedofsound.util.BluetoothDevicePreference
import java.util.*

/**
 * Sound service activation manager. Used to start the service at boot;
 * referenced in the manifest.
 */
class SoundServiceManager : BroadcastReceiver() {

    /**
     * Keep track of the bluetooth state here, as the undocumented broadcasts
     * might not be sticky. Not sure if the official ones are either.
     */
    private var bluetoothConnected = false

    /**
     * Receive a broadcast and start the service or update the tracking state.
     */
    override fun onReceive(context: Context, intent: Intent) {
        val action = intent.action
        Log.d(TAG, "Received intent with action $action")

        // resume tracking if we're also in a satisfactory mode
        if (action == Intent.ACTION_POWER_CONNECTED ||
            action == Intent.ACTION_POWER_DISCONNECTED ||
            action == Intent.ACTION_HEADSET_PLUG) {
            setTracking(context, this.shouldTrack(context))
        } else if (action == BluetoothDevice.ACTION_ACL_CONNECTED ||
            action == BluetoothDevice.ACTION_ACL_DISCONNECTED) {
            Log.d(TAG, "Bluetooth ACL connect/disconnect event")

            // check whether we care about this event at all
            val prefs = PreferenceManager.getDefaultSharedPreferences(context)
            if (!prefs.getBoolean("enable_bluetooth", false)) {
                Log.v(TAG, "Bluetooth pref disabled; ignoring")
                return
            }

            // grab the device address and check it against our list of things
            val shouldCare = when (val device = intent.getParcelableExtra<BluetoothDevice>(BluetoothDevice.EXTRA_DEVICE)) {
                null -> false
                else -> isSelectedBluetoothDevice(prefs, device.address)
            }
            if (!shouldCare) {
                Log.v(TAG, "Bluetooth pref enabled but we don't care about this device")
                return
            }

            // set the bluetooth state
            this.bluetoothConnected = action == BluetoothDevice.ACTION_ACL_CONNECTED
            Log.v(TAG, "Bluetooth active: ${this.bluetoothConnected}")

            // start or stop tracking
            setTracking(context, this.shouldTrack(context))
        } else if (action == LOCALE_FIRE) {
            // Tasker!
            val bundle = intent.getBundleExtra(LOCALE_BUNDLE)
            if (bundle != null) {
                val state = bundle.getBoolean(SoundService.SET_TRACKING_STATE, true)
                setTracking(context, state)
            }
        } else {
            // external intent (likely from notification "Stop" action)
            val state = intent.extras?.getBoolean(SoundService.SET_TRACKING_STATE)
            if (state != null) {
                setTracking(context, state)
            }
        }
    }

    /**
     * Determine whether we should be tracking.
     *
     * @param context Application context
     * @return suggested state
     */
    private fun shouldTrack(context: Context): Boolean {
        // load preferences
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        val powerPreference = prefs.getBoolean("enable_only_charging", false)
        val bluetoothPreference = prefs.getBoolean("enable_bluetooth", false)

        // get power status
        val plugFilter = IntentFilter(Intent.ACTION_BATTERY_CHANGED)
        val powerStatus = context.applicationContext.registerReceiver(null, plugFilter)
        var powerConnected = false
        if (powerStatus != null) {
            val plugState = powerStatus.getIntExtra(BatteryManager.EXTRA_PLUGGED, -1)
            powerConnected = plugState == BatteryManager.BATTERY_PLUGGED_AC || plugState == BatteryManager.BATTERY_PLUGGED_USB
        } else {
            Log.e(TAG, "Power status was null")
        }

        // don't track if power is disconnected and we care
        if (powerPreference && !powerConnected) {
            Log.v(TAG, "Power preference active & disconnected")
            return false
        }

        // also activate if bluetooth is connected
        if (bluetoothPreference && this.bluetoothConnected) {
            Log.v(TAG, "Bluetooth connected")
            return true
        }

        // anything else is a no-go
        return false
    }

    /**
     * Return whether the given Bluetooth address is enabled for tracking.
     *
     * Loaded from user preferences.
     */
    private fun isSelectedBluetoothDevice(prefs: SharedPreferences, address: String): Boolean {
        // fetched saved devices
        val addresses = prefs.getStringSet(BluetoothDevicePreference.KEY, HashSet())

        // no selected devices means that *any* bluetooth device is valid
        return if (addresses!!.size == 0) {
            true
        } else addresses.contains(address)

    }

    companion object {
        private const val TAG = "SoundServiceManager"

        const val LOCALE_BUNDLE = "com.twofortyfouram.locale.intent.extra.BUNDLE"
        const val LOCALE_BLURB = "com.twofortyfouram.locale.intent.extra.BLURB"
        const val LOCALE_FIRE = "com.twofortyfouram.locale.intent.action.FIRE_SETTING"

        /**
         * Set the tracking state by sending a service start command or broadcast.
         */
        private fun setTracking(context: Context, desiredState: Boolean) {
            Log.d(TAG, "Setting tracking desiredState: $desiredState")

            // only call startForegroundService if we want to start tracking;
            // in 8.0+ the service *must* start up and present a notification
            // within 5 seconds or Android will kill the whole app
            if (desiredState) {
                val serviceIntent = Intent(context, SoundService::class.java)
                serviceIntent.putExtra(SoundService.SET_TRACKING_STATE, desiredState)
                ContextCompat.startForegroundService(context, serviceIntent)
            } else {
                val commandIntent = Intent(SoundService.SET_TRACKING_STATE)
                commandIntent.putExtra(SoundService.SET_TRACKING_STATE, desiredState)
                LocalBroadcastManager.getInstance(context).sendBroadcast(commandIntent)
            }

        }
    }

}
